src/ciphersuite/mod.rs

```diff
@@ -201,84 +202,223 @@ impl KdfLabel {
+            value: Vec::new(),
+            mls_version: ProtocolVersion::default(),
+        }
+    }
+}
+
+impl PartialEq for Secret {
+    // Constant time comparison.
+    fn eq(&self, other: &Secret) -> bool {
+        // These values can be considered public and checked before the actual
+        // comparison.
+        if self.ciphersuite != other.ciphersuite
+            || self.mls_version != other.mls_version
+            || self.value.len() != other.value.len()
+        {
+            log::error!("Incompatible secrets");
+            log::trace!(
+                "  {} {} {}",
+                self.ciphersuite.name,
+                self.mls_version,
+                self.value.len()
+            );
+            log::trace!(
+                "  {} {} {}",
+                other.ciphersuite.name,
+                other.mls_version,
+                other.value.len()
+            );
+            return false;
+        }
+        let mut diff = 0u8;
+        for (l, r) in self.value.iter().zip(other.value.iter()) {
+            diff |= l ^ r;
+        }
+        diff == 0
+    }
 }
 
 impl Secret {
     /// Randomly sample a fresh `Secret`.
-    pub(crate) fn random(length: usize) -> Self {
+    /// This default random initialiser uses the default Secret length of `hash_length`.
+    pub(crate) fn random(
+        ciphersuite: &'static Ciphersuite,
+        version: impl Into<Option<ProtocolVersion>>,
+    ) -> Self {
+        let mls_version = version.into().unwrap_or_default();
+        log::trace!(
+            "Creating a new random secret for {:?} and {:?}",
+            ciphersuite.name,
+            mls_version
+        );
+        // if ciphersuite.name == CiphersuiteName::MLS10_128_DHKEMP256_AES128GCM_SHA256_P256 {
+        //     log::trace!("{:?}", backtrace::Backtrace::new());
+        // }
         Secret {
-            value: get_random_vec(length),
+            value: get_random_vec(ciphersuite.hash_length()),
+            mls_version,
+            ciphersuite,
+        }
+    }
+
+    /// Create an all zero secret.
+    pub(crate) fn zero(ciphersuite: &'static Ciphersuite, mls_version: ProtocolVersion) -> Self {
+        Self {
+            value: vec![0u8; ciphersuite.hash_length()],
+            mls_version,
+            ciphersuite,
         }
     }
 
+    /// Create a new secret from a byte vector.
+    pub(crate) fn from_slice(
+        bytes: &[u8],
+        mls_version: ProtocolVersion,
+        ciphersuite: &'static Ciphersuite,
+    ) -> Self {
+        Secret {
+            value: bytes.to_vec(),
+            mls_version,
+            ciphersuite,
+        }
+    }
+
+    /// HMAC-Hash(salt, IKM). For all supported ciphersuites this is the same
+    /// HMAC that is also used in HKDF.
+    /// Compute the HMAC on `self` with key `ikm`.
+    pub(crate) fn mac(&self, ikm: &Secret) -> Self {
+        // FIXME: this should be on ikm and the other input shouldn't be a secret.
+        self.hkdf_extract(ikm)
```

Comment:
I believe that this is how it was before, but isn't it generally a bit strange that `hkdf_extract` and not `hmac`?

Reply:
It's the same thing. This code was only moved around.